Add regex for shared object file names

sharidas opened the pull request:
Add regular expression for shared object file name.
The file name can be for example 'a.so' or 'a.so.0.0.1'.
So this regular expression would help us to figure
out that file has 'so' extension too.

Signed-off-by: Sujith H <sharidasan@owncloud.com>

<!--
Thanks for submitting a change to ownCloud!

This is the bug tracker for the Server component. Find other components at https://github.com/owncloud/core/blob/master/.github/CONTRIBUTING.md#guidelines

For fixing potential security issues please see https://owncloud.org/security/

To make it possible for us to get your change reviewed and merged please carefully fill out the requested information below.

Please note that any kind of change needs first be submitted to the master branch which holds the next major version of ownCloud.

We will carefully discuss if your change can or has to be backported to stable branches.

Please set the following labels:

- Set label "3 - To review" for review or "2 - Development" in case the PR still has open tasks
- Set label "backport-request" if backport is needed
- Assignment: assign to self
- Milestone: set the same as the ticket this PR fixes, or "development" by default
- Reviewers: pick at least one
-->

## Description
<!--- Describe your changes in detail -->
Adding regular expression to check if the file is of extension `filename.so.0.0.1`. This regular expression tries to check if there is a `.so` in the file name. And if so, it would use the extension `.so`. We cannot rely completely on `mimetypemapping.dist.json`, where it expects file with `filename.so` only.

## Related Issue
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->
- Fixes <issue_link>

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
In this pr we try to solve the problem of shared object filenames which have multiple `.` and with the current mechanism, it would not be able to detect the extension for `filename.so.0.0.0.0.0.1`. With this pr the extension of `filenam.so.0.0.0.0.1` is taken as `.so`.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
- Tested in the UI by uploading a file with a `.so` extension only and `.so.0.0.0.1`. Both were caught when firewall rule was added.

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests

## Checklist:
<!-- Tick the checkboxes when done. -->
<!-- Raise documentation ticket in https://github.com/owncloud/documentation -->
- [ ] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 

## Open tasks:
<!-- In case of incomplete PR, please list the open tasks here -->
- [ ] Backport (if applicable set "backport-request" label and remove when the backport was done)

sharidas committed:
Add regex for shared object file names

Add regular expression for shared object file name.
The file name can be for example 'a.so' or 'a.so.0.0.1'.
So this regular expression would help us to figure
out that file has 'so' extension too.

Signed-off-by: Sujith H <sharidasan@owncloud.com>

PVince81 commented on lib/private/Files/Type/Detection.php:
```diff
@@ -170,9 +170,19 @@ public function detectPath($path) {
+			 * lib.so.0.0.1 ( shared object file ) and if not then
+			 * use the else logic.
+			 */
+			\preg_match("/(.*?)\.(so)(.*)/i", $fileName, $match);
```
mind isolating this into a private function ?

PVince81 commented on lib/private/Files/Type/Detection.php:
```diff
@@ -150,6 +150,18 @@ private function loadMappings() {
+	 */
+	private function detectSharedObjFileName($fileName) {
+		\preg_match("/(.*?)\.(so)\.(.*)/i", $fileName, $match);
+		return (\count($match) === 4) && ($match[2] === 'so');
```
is it always 4 components or can it be less ?

```
lrwxrwxrwx 1 root root   20 Aug 20 21:31 /usr/lib64/libSDL-1.2.so.0 -> libSDL-1.2.so.0.11.4
-rwxr-xr-x 1 root root 431K Aug 20 21:31 /usr/lib64/libSDL-1.2.so.0.11.4
lrwxrwxrwx 1 root root   20 Nov 12 16:20 /usr/lib64/libSDL2-2.0.so.0 -> libSDL2-2.0.so.0.9.0
-rwxr-xr-x 1 root root 1.3M Nov 12 16:20 /usr/lib64/libSDL2-2.0.so.0.9.0
lrwxrwxrwx 1 root root   20 Aug 27 17:48 /usr/lib64/libSDL_gfx.so.15 -> libSDL_gfx.so.15.9.1
-rwxr-xr-x 1 root root  89K Aug 27 17:48 /usr/lib64/libSDL_gfx.so.15.9.1
lrwxrwxrwx 1 root root   25 Jun 10  2018 /usr/lib64/libSDL_image-1.2.so.0 -> libSDL_image-1.2.so.0.8.4
-rwxr-xr-x 1 root root  59K Jun 10  2018 /usr/lib64/libSDL_image-1.2.so.0.8.4
```

sharidas replied:
If the match happens for the .so file, the `$match` will have 4 components. Else it will have zero.
So for example if the fileName is `testso.1.2`, the `$match` will have zero components. If the fileName is `test.so.1.2`, then `$match` will be `[0]=> string(11) "test.so.1.2" [1]=> string(4) "test" [2]=> string(3) ".so" [3]=> string(3) "1.2" `

sharidas commented:
Current status of the PR:
- Did an update to the regex. The update was done so that after `.so` it may have `.` followed by anything. Or the file name ends with `.so`.

PVince81 commented on lib/private/Files/Type/Detection.php:
```diff
@@ -150,6 +150,18 @@ private function loadMappings() {
+	 * @return bool true if file is a shared object else false
+	 */
+	private function detectSharedObjFileName($fileName) {
+		\preg_match("/(.*?)(\.so)[\.]*(.*)/i", $fileName, $match);
```
will this also match ".so." and ".so.x" ?

for the excluded cases please add test cases in DetectionTest to confirm they are not matched

sharidas replied:
The updated PR would match the cases requested. Updated the test with cases that does not match the regex.

DeepDiver1975 commented:
MacOS has .dylib extention - are we handling this file extension yet? - just popped up ...

sharidas commented:
> MacOS has .dylib extention - are we handling this file extension yet? - just popped up ...

Aah thanks for pointing it out. IMHO it does make sense to consider the missing `.dylib` extension.

sharidas committed:
Add mime type for shared lib extension dylib (mac)

Add mime type for shared lib extnesion dylib.
dylib extension is required for MacOS.

Signed-off-by: Sujith H <sharidasan@owncloud.com>

sharidas commented:
Current state of the PR:
- Updated the PR to include `.dylib` extension by adding it to `resources/config/mimetypemapping.dist.json`
- Updated the regex to match the query posted https://github.com/owncloud/core/pull/33894#discussion_r242898430
- Updated the test to check the cases where the regex does not match.
- Added a doc in the code, it would help the future developers while going through the Detection code.

Hence I would appreciate a review for the updated changes. @PVince81

PVince81 reviewed (approved):
:+1:

PVince81 commented:
@sharidas please backport for 10.1.1

sharidas commented:
Backport PR to stable10 https://github.com/owncloud/core/pull/34082
